node/actors/bft/src/replica/tests.rs

```diff
@@ -1,54 +1,477 @@
+        .own_key()
+        .sign_msg(ConsensusMsg::LeaderPrepare(leader_prepare.clone()));
+
+    let res = util.dispatch_leader_prepare(leader_prepare_signed).await;
+    assert_matches!(
+        res,
+        Err(LeaderPrepareError::ProposalOversizedPayload{ payload_size, header }) => {
+            assert_eq!(payload_size, payload_oversize);
+            assert_eq!(header, leader_prepare.proposal);
+        }
+    )
+}
+
+#[tokio::test]
+async fn leader_prepare_proposal_mismatched_payload() {
+    let mut util = UTHarness::new().await;
+
+    let mut leader_prepare = util
+        .new_procedural_leader_prepare()
+        .await
+        .cast::<LeaderPrepare>()
+        .unwrap()
+        .msg;
+    leader_prepare.proposal_payload = Some(util.rng().gen());
+    let leader_prepare_signed = util
+        .own_key()
+        .sign_msg(ConsensusMsg::LeaderPrepare(leader_prepare.clone()));
+
+    let res = util.dispatch_leader_prepare(leader_prepare_signed).await;
+    assert_matches!(res, Err(LeaderPrepareError::ProposalMismatchedPayload))
+}
+
+#[tokio::test]
+async fn leader_prepare_proposal_when_previous_not_finalized() {
+    let mut util = UTHarness::new().await;
+
+    let replica_prepare = util.new_current_replica_prepare(|_| {});
+    util.dispatch_replica_prepare(replica_prepare.clone())
+        .unwrap();
+
+    let mut leader_prepare = util
+        .recv_signed()
+        .await
+        .unwrap()
+        .cast::<LeaderPrepare>()
+        .unwrap()
+        .msg;
+
+    let high_vote = util.rng().gen();
+    leader_prepare.justification = util.new_prepare_qc(|msg| msg.high_vote = high_vote);
+
+    let leader_prepare_signed = util
+        .own_key()
+        .sign_msg(ConsensusMsg::LeaderPrepare(leader_prepare.clone()));
+
+    let res = util.dispatch_leader_prepare(leader_prepare_signed).await;
+    assert_matches!(
+        res,
+        Err(LeaderPrepareError::ProposalWhenPreviousNotFinalized)
+    )
+}
+
+#[tokio::test]
+async fn leader_prepare_proposal_invalid_parent_hash() {
+    let mut util = UTHarness::new().await;
+
+    let replica_prepare_signed = util.new_current_replica_prepare(|_| {});
+    let replica_prepare = replica_prepare_signed
+        .clone()
+        .cast::<ReplicaPrepare>()
+        .unwrap()
+        .msg;
+    util.dispatch_replica_prepare(replica_prepare_signed.clone())
+        .unwrap();
+    let mut leader_prepare = util
+        .recv_signed()
+        .await
+        .unwrap()
+        .cast::<LeaderPrepare>()
+        .unwrap()
+        .msg;
+
+    let junk: BlockHeaderHash = util.rng().gen();
+    leader_prepare.proposal.parent = junk;
+    let leader_prepare_signed = util
+        .own_key()
+        .sign_msg(ConsensusMsg::LeaderPrepare(leader_prepare.clone()));
+
+    let res = util.dispatch_leader_prepare(leader_prepare_signed).await;
+    assert_matches!(
+        res,
+        Err(LeaderPrepareError::ProposalInvalidParentHash { correct_parent_hash, received_parent_hash, header }) => {
```

Comment:
Nit: Rust formatter is quite bad at formatting code inside macros. This overly long line could be formatted as

```rust
assert_matches!(
    res,
    Err(LeaderPrepareError::ProposalInvalidParentHash {
        correct_parent_hash,
        received_parent_hash,
        header,
    }) => {
        assert_eq!(correct_parent_hash, replica_prepare.high_vote.proposal.hash());
        assert_eq!(received_parent_hash, junk);
        assert_eq!(header, leader_prepare.proposal);
    }
);
```

Reply:
Fixed.